File: leetcode/easy/993-cousinsInBinaryTree.js
/*
In a binary tree, the root node is at depth 0, and children of each depth k node are at depth k+1.

Two nodes of a binary tree are cousins if they have the same depth, but have different parents.

We are given the root of a binary tree with unique values, and the values x and y of two different nodes in the tree.

Return true if and only if the nodes corresponding to the values x and y are cousins.

Example 1:
Input: root = [1,2,3,4], x = 4, y = 3
Output: false

Example 2:
Input: root = [1,2,3,null,4,null,5], x = 5, y = 4
Output: true
*/

// first pass solution
var isCousins = function(root, x, y) {
    let dict = { x: null, y: null };
    
    function traverse(parent, root, depth) {
        if (root.val === x) dict[x] = [parent, depth];
        else if (root.val === y) dict[y] = [parent, depth];
        
        if (root.left) traverse(root.val, root.left, depth + 1);    
        if (root.right) traverse(root.val, root.right, depth + 1);
    }
    
    traverse(null, root, 0);
    
    return dict[x][0] !== dict[y][0] && dict[x][1] === dict[y][1];
};

File: leetcode/easy/993-cousinsInBinaryTree.java
// BFS
class Solution {
    public boolean isCousins(TreeNode root, int x, int y) {
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        
        while (!queue.isEmpty()) {
            int len = queue.size();
            TreeNode xParent = null;
            TreeNode yParent = null;
            
            for (int i = 0; i < len; i++) {
                TreeNode curr = queue.poll();
                
                if (curr.left != null) {
                    if (curr.left.val == x) xParent = curr;
                    else if (curr.left.val == y) yParent = curr;
                    queue.add(curr.left);
                }
                if (curr.right != null) {
                    if (curr.right.val == y) yParent = curr;
                    else if (curr.right.val == x) xParent = curr; 
                    queue.add(curr.right);
                }
            }
            if (xParent != null && yParent != null) {
                if (xParent != yParent) return true;
                else break;
            }
        }
        return false;
    }
}

// DFS
class Solution {
    public boolean isCousins(TreeNode root, int x, int y) {
        if (root == null) return false;
        
        Map<Integer, int[]> map = new HashMap<>();
        
        DFS(root, x, y, map, 0);
        
        if (map.containsKey(x) && map.containsKey(y)) {
            return map.get(x)[0] != map.get(y)[0] && map.get(x)[1] == map.get(y)[1];
        }
        return false;
    }
    
    public void DFS(TreeNode root, int x, int y, Map<Integer, int[]> map, int depth) {
        if (root.left != null) {
            if (root.left.val == x) map.put(x, new int[] { root.val, depth });
            else if (root.left.val == y) map.put(y, new int[] { root.val, depth });
            DFS(root.left, x, y, map, depth + 1);
        }
        
        if (root.right != null) {
            if (root.right.val == x) map.put(x, new int[] { root.val, depth });
            else if (root.right.val == y) map.put(y, new int[] { root.val, depth });
            DFS(root.right, x, y, map, depth + 1);
        }
    }
}
